### PR TITLE
crimson/os/seastore: consider segment_header_t::modify_time as the segments' modify_time for no-tail OOL segments

### PR DESCRIPTION
**Pull request by xxhdx1985126**
OOL segments don't contain record headers anymore, we use segment_header_t::modify_time as the approximation of segment modify_time

Fixes: https://tracker.ceph.com/issues/67106
Signed-off-by: Xuehan Xu <xuxuehan@qianxin.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/crimson/os/seastore/async_cleaner.cc [hunk 1]
```diff
@@ -1372,16 +1372,20 @@ SegmentCleaner::mount_ret SegmentCleaner::mount()
...
+        sea_time_point modify_time = mod_to_timepoint(
+	  tail_valid ? tail.modify_time : header.modify_time);
+        std::size_t num_extents = tail_valid ? tail.num_extents : 0;
         if ((modify_time == NULL_TIME && num_extents == 0) ||
-            (modify_time != NULL_TIME && num_extents != 0)) {
-          segments.update_modify_time(segment_id, modify_time, num_extents);
+	    modify_time != NULL_TIME) {
+	  DEBUG("updating modify time for segment {}, mod time {}, num_extents {}",
+	    segment_id, modify_time, num_extents);
+          segments.init_modify_time(segment_id, modify_time, num_extents);
```
[the remark this answers is not in the document]
Comment (xxhdx1985126) [reply src/crimson/os/seastore/async_cleaner.cc:1388]: Yeah, will do

File: src/crimson/os/seastore/async_cleaner.h [hunk 2]
```diff
@@ -75,6 +75,14 @@ struct segment_info_t {
 
   void set_closed();
 
+  void init_modify_time(sea_time_point _modify_time, std::size_t _num_extents) {
+    ceph_assert(modify_time == NULL_TIME);
+    ceph_assert(num_extents == 0);
+    ceph_assert(_modify_time != NULL_TIME);
+    modify_time = _modify_time;
+    num_extents = _num_extents;
```
[the remark this answers is not in the document]
Comment (xxhdx1985126) [reply src/crimson/os/seastore/async_cleaner.h:83]: Yeah:-)

File: src/crimson/os/seastore/async_cleaner.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (xxhdx1985126) [reply src/crimson/os/seastore/async_cleaner.cc]: Done, please take a look again, thanks:-)

File: src/crimson/os/seastore/seastore_types.h [hunk 3]
```diff
@@ -1347,8 +1347,6 @@ std::ostream &operator<<(std::ostream &out, mod_time_point_printer_t tp);
 constexpr sea_time_point
 get_average_time(const sea_time_point& t1, std::size_t n1,
                  const sea_time_point& t2, std::size_t n2) {
-  assert(t1 != NULL_TIME);
-  assert(t2 != NULL_TIME);
```
Comment (cyx1231st) [inline src/crimson/os/seastore/seastore_types.h:1351]: These are still valid and shouldn't be removed, right?

File: src/crimson/os/seastore/async_cleaner.h [hunk 4]
```diff
@@ -75,6 +76,14 @@ struct segment_info_t {
 
   void set_closed();
 
+  void init_modify_time(sea_time_point _modify_time, std::size_t _num_extents) {
```
Comment (cyx1231st) [inline src/crimson/os/seastore/async_cleaner.h:79]: Let's drop `_num_extents` which is not used, keeping `ceph_assert(num_extents == 0);` should be good.

File: src/crimson/os/seastore/seastore_types.h (hunk 3 above)
Comment (xxhdx1985126) [reply src/crimson/os/seastore/seastore_types.h:1351]: I called `get_average_time()` and `segments_info_t::update_modify_time()` no matter whether `num_extents` is zero or not, which led to violation of these asserts.

Comment (xxhdx1985126) [reply src/crimson/os/seastore/seastore_types.h:1351]: Maybe `get_averge_time()` and `segments_info_t::update_modify_time()` only needs to be called in cases when `num_extents` is non-zero, am I right?

Comment (cyx1231st) [reply src/crimson/os/seastore/seastore_types.h:1351]: The original asserts were working, and looks `segments_info_t::update_modify_time()` was already correct and reasonable.

For async_cleaner.cc 1388, How about:
```c++
if (num_extents == 0) {
  ceph_assert(tail_time == NULL_TIME);
  // pass
} else {
  ceph_assert(header_time != NULL_TIME);
  ceph_assert(tail_time != NULL_TIME);
  sea_time_point avg_time = get_average_time(
	    header_time, 1, tail_time, 1);
  segments.update_modify_time(segment_id, avg_time, num_extents);
}
```

Comment (cyx1231st) [reply src/crimson/os/seastore/seastore_types.h:1351]: Anyway, it's meaningless to get average from a NULL_TIME.

Comment (xxhdx1985126) [reply src/crimson/os/seastore/seastore_types.h:1351]: Done:-)

Comment (cyx1231st) [reply src/crimson/os/seastore/seastore_types.h:1351]: @xxhdx1985126 forget to remove them? Others LGTM.

Comment (xxhdx1985126) [reply src/crimson/os/seastore/seastore_types.h]: Ah, yes, sorry.

**Comment by cyx1231st** [comment]
jenkins retest this please

**Comment by cyx1231st** [comment]
https://shaman.ceph.com/builds/ceph/wip-yingxin-testing-2024-08-07-1253/
https://pulpito.ceph.com/yingxin-2024-08-07_07:12:11-crimson-rados-wip-yingxin-testing-2024-08-07-1253-distro-default-smithi/
* 7841889/90: https://tracker.ceph.com/issues/63844
* 7841905/06: https://tracker.ceph.com/issues/66761

**Comment by cyx1231st** [comment]
jenkins test make check

**Pull request merged**